### PR TITLE
Add rainbow effect to led strip and cleanup

### PR DESCRIPTION
**Pull request by ASDosjani**
Betaflight PR: https://github.com/betaflight/betaflight/pull/12323

Demonstration: https://youtu.be/i4qeOpRaKrM
New cli commands:
`ledstrip_rainbow_delta` - delta between the leds in hsv color
`ledstrip_rainbow_freq` - frequency of hsv change

Try to change these values with the `set` command.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by McGiverGim** [comment]
Don't modify translations files. Only English. The others will be modified from Crowdin automatically, and the PR is simpler with less modified files.

File: locales/en/messages.json
Comment (McGiverGim) [inline locales/en/messages.json:3495]: Any change in the messages files will force the translators to traslate them again. I don't use leds, but I suppose the change is because it describes better what is it. Am I right?
I recommend for any string added/modified adding a description that will be seen for translators:
```json
"ledStripLarsonOverlay": {
        "message": "Larson scanner",
        "description": "One of the modes for the LED strip feature"
},
```

File: locales/en/messages.json
Comment (McGiverGim) [inline locales/en/messages.json:3498]: Why is this deleted? I suppose now is the new `ledStripModifierTitle`?

File: locales/en/messages.json
Comment (McGiverGim) [inline locales/en/messages.json:3504]: Add the same description tag that commented before, please.

File: src/js/injected_methods.js
Comment (McGiverGim) [inline src/js/injected_methods.js:66]: Why is this not working? This only happens in Betaflight or there is some info out about this?

File: src/js/msp/MSPHelper.js
Comment (McGiverGim) [inline src/js/msp/MSPHelper.js:1175]: What happens if LED64 is not enabled? This works or crashes?

**Review by McGiverGim — CHANGES_REQUESTED** [review]
Some comments, I don't have LED to test it, so I can't confirm that this works/don't breaks anything.

File: locales/en/messages.json
Comment (ASDosjani) [reply locales/en/messages.json:3495]: Yeah, it describes the function better, I think they leaved this translation key because legacy support.

File: locales/en/messages.json
Comment (ASDosjani) [reply locales/en/messages.json:3498]: It's not used anymore.
Before:
Screenshot_2023-02-10-13-28-35-140-edit_com betaflight betaflight_configurator_debug

After:
Screenshot_2023-02-10-13-27-45-971-edit_com betaflight betaflight_configurator_debug

File: src/js/injected_methods.js
Comment (ASDosjani) [reply src/js/injected_methods.js:66]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

I don't wanted to use BigInt because I would have to duplicate some functions then.

File: src/js/msp/MSPHelper.js
Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:1175]: It's not related to led count, both works. The 64 bit is the led configuration size.
This is the reason why I had to change it to 64 bit:
https://github.com/betaflight/betaflight/pull/12323#issue-1572957564

File: src/js/injected_methods.js
Comment (McGiverGim) [reply src/js/injected_methods.js:66]: Ok, it's simply the size of the data, not any kind of problem or bug. I don't know if have a "false" function for that is good, or simply call twice the readU32 directly when reading the MSP command as we do for other similar data (I think remember the flight modes maybe?). I think is more clear. @haslinghuis what do you think?

File: src/js/msp/MSPHelper.js
Comment (McGiverGim) [reply src/js/msp/MSPHelper.js:1175]: But will work too with older firmware?

Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:1175]: Probably not, I should bind it to firmware version.

Comment (McGiverGim) [reply src/js/msp/MSPHelper.js:1175]: This is easier to do with two 32 bits fields. One for led_low, and the other for led_high that will be zero for older version and only read for new version.

**Comment by ASDosjani** [comment]
@McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0.
Please review it if you have time.

File: locales/en/messages.json
Comment (ASDosjani) [reply locales/en/messages.json]: Wdym?

Comment (ASDosjani) [reply locales/en/messages.json]: > Add the same description tag that commented before, please.

I think it's pretty obvious what this string for, don't know what should I write in the description.

Comment (McGiverGim) [reply locales/en/messages.json]: If you see only ”Rainbow" without any other context you know what is it for? That is what see the translators. They don't know this is for LED feature. For this reason I suggested in the other change something generic like "One of the modes for the LED strip feature". This helps a lot the translators to know where the string will be used.

File: src/js/msp/MSPHelper.js
Comment (ASDosjani) [reply src/js/msp/MSPHelper.js]: I think this 64 bit function can be useful in the future.

File: locales/en/messages.json
Comment (ASDosjani) [reply locales/en/messages.json]: Done, I added description in the English file.

**Review by McGiverGim — CHANGES_REQUESTED** [review]
I think you have modified the translations files by error again. They will be modified from Crowdin  when this PR is merged.

**Comment by McGiverGim** [comment]
> @McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0. Please review it if you have time.

The modifications to translations files will be lost. We don't upload them to Crowdin, and we overwrite them with the contents of Crowdin. For this reason we ask to not modify them to the developers. Only the English file, is the only one that we upload to Crowdin.

**Comment by ASDosjani** [comment]
> I think you have modified the translations files by error again. They will be modified from Crowdin when this PR is merged.

Does Crowdin know that I changed the name of a key and not adding an extra that translators should translate again?
I think it's easier to fix this in this PR.

**Comment by ASDosjani** [comment]
> > @McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0. Please review it if you have time.
> 
> The modifications to translations files will be lost. We don't upload them to Crowdin, and we overwrite them with the contents of Crowdin. For this reason we ask to not modify them to the developers. Only the English file, is the only one that we upload to Crowdin.

Fixed.

File: src/js/injected_methods.js
Comment (haslinghuis) [reply src/js/injected_methods.js:66]: Agree with using a 2 x 32 bit struct to complement the 32 and 64 led_strip.

Comment (ASDosjani) [reply src/js/injected_methods.js:66]: > Agree with using a 2 x 32 bit struct to complement the ledstrip_32 and ledstrip_64 feature.

So is it good now or should I change it?
I don't understand exactly.

Comment (haslinghuis) [reply src/js/injected_methods.js:66]: Please use 32 bit data structure. With ledstrip 64 enabled we can use a second 32 bit structure.
Hardest part is remaining backwards compatibility with previous versions.

Comment (ASDosjani) [reply src/js/injected_methods.js:66]: > Please use 32 bit data structure. With ledstrip 64 enabled we can use a second 32 bit structure. Hardest part is remaining backwards compatibility with previous versions.

I think you misunderstand the problem. It's not related to led strip length (32 or 64), instead it's the led config length, each configured led has this which contains it's configuration. Previously it was 32bit long but with the new feature I couldn't fit into this so I had to expand it to 64 bit.

The problem is that JavaScript can't use 64 bit numbers only 53 bit so I call the the readU32 twice and return them in an array. I added readU64 where this is happening. It's for clearity and code readability.
I think @McGiverGim wants me not to implement this readU64 and use twice readU32 directly in the code.

I already fixed the backward compatible problem.

Comment (haslinghuis) [reply src/js/injected_methods.js:66]: Agree with @McGiverGim Please use twice U32

Comment (ASDosjani) [reply src/js/injected_methods.js:66]: Alright, I change it.
Limon added the 64 led feature to cloud build.

File: src/css/tabs/led_strip.less
Comment (haslinghuis) [inline src/css/tabs/led_strip.less:154]: Needed for backward compatibility

File: src/js/msp/MSPHelper.js
Comment (haslinghuis) [inline src/js/msp/MSPHelper.js:23]: Needed for backward compatibility

File: src/js/msp/MSPHelper.js
Comment (haslinghuis) [inline src/js/msp/MSPHelper.js:2538]: Is this a bug?

File: src/js/msp/MSPHelper.js
Comment (haslinghuis) [inline src/js/msp/MSPHelper.js:2543]: 23 instead of 22, a bug?

Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:2543]: No, these changes are intended. The overlay is expanded from 6bit to 7bit and I had to shift every offset after them. That's why I asked if I should allocate more like 8-9 bits for the overlay and function.
I will update the PR with an illustration and then you will understand how it works.

File: src/js/msp/MSPHelper.js
Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:23]: I handle it here:
image

File: src/css/tabs/led_strip.less
Comment (ASDosjani) [reply src/css/tabs/led_strip.less:154]: I think we don't need this, the configurator only compatible from bf 4.0+ and I just checked and bf 4.0 doesn't contain this overlay.
It's even hidden in the process_html function in `led_strip.js` line 88 and only occurs once in the file, there.
So I think it's not supported now anyway.

File: src/js/msp/MSPHelper.js
Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:2543]: @haslinghuis wait a bit, I might be stupid. I have to review my code. I might used wrong numbers.

Comment (ASDosjani) [reply src/js/msp/MSPHelper.js:2550]: @haslinghuis see in the MSPHelper.js file line 1169.

File: src/js/msp/MSPHelper.js
Comment (haslinghuis) [inline src/js/msp/MSPHelper.js:1263]: Reduce Indentation :)

**Review by nerdCopter — APPROVED** [review]
* i've seen video & reports of it working, i cannot test personally, but based on feedback i'll add my approval.

**Comment by nerdCopter** [comment]
* please marked any items as resolved if they are indeed resolved.

**Review by haslinghuis — APPROVED** [review]

**Pull request merged**
